appvia/kore
# Add immutable cluster and team identities and tag all resources

## Summary

This introduces a concept of an immutable identity to both teams and clusters. These identifiers have a lifecycle outside the objects they represent, meaning that even if a cluster is deleted, we can continue to track its ownership for (e.g.) cost allocation purposes.

It also ensures that all EKS, AKS and GKE resources are tagged with both kore-team and kore-cluster tags containing these identifiers.

**Which issue(s) this PR resolves**:
Addresses #1076
Addresses #1078 
Fixes #1141 

## Checklist

 - [ ] I've created a PR to update the [documentation](https://github.com/appvia/kore-docs): n/a

## Changes

* Introduce asset identity generator to the kore interface.
* Persist generated asset identities to the DB layer when they are assigned.
* Ensure that clusters have team and cluster identity labels when created.
* Ensure all teams and clusters have identifiers assigned on startup, to ensure backwards compatibility.
* Ensure teams are assigned an identifier on creation.
* Validate supplied asset and team identifiers when managing clusters - allowing teams to re-use a previously deleted asset identifier if they (for example) replace a cluster.
* Tag all EKS, AKS and GKE resources with the `kore-cluster` and `kore-team` tags containing these identifiers.

### Additional changes

* Fix #1141 by using an empty object instead of the supplied object to get the original object when updating CRDs in the Store Client.
* Add missing static data for Azure and update UI E2E test to use that so the UI E2E should pass whether or not the Azure cloudinfo data is available. 

## Testing
Testing completed:
* Create a team, run `kore get team teamname -o yaml` and see the labels in the metadata:
```
apiVersion: org.kore.appvia.io/v1
kind: Team
metadata:
  creationTimestamp: "2020-07-22T12:47:56Z"
  labels:
    kore.appvia.io/teamid: bsc3av58lboip6rtho80
  name: mark
  namespace: kore
spec:
  description: mark
  summary: mark
status:
  conditions: []
  status: Success
```

* Create a cluster, run `kore get cluster -t teamname clustername -o yaml` and see the labels in the metadata:

```
apiVersion: clusters.compute.kore.appvia.io/v1
kind: Cluster
metadata:
  creationTimestamp: "2020-07-22T12:48:10Z"
  finalizers:
  - cluster.clusters.kore.appvia.io
  generation: 2
  labels:
    kore.appvia.io/clusterid: bscrq258lbop88s8t9pg
    kore.appvia.io/teamid: bsc3av58lboip6rtho80
  name: mark-gke-test
  namespace: mark
  resourceVersion: "1771200"
  selfLink: /apis/clusters.compute.kore.appvia.io/v1/namespaces/mark/clusters/mark-gke-test
  uid: 53fdb73c-a241-4bf3-9c0b-845d37a51e24
spec:
  configuration:
    authProxyAllowedIPs:
    ...
```

* Checked in database to see that these identifiers arepersisted in the dedicated tables for team identifiers and team assets.
* Created AKS, EKS and GKE clusters, observed tagging of the relevant resources in all three providers.

![image](https://user-images.githubusercontent.com/7505593/88399542-f2404500-cdbe-11ea-93d5-e1a63c2266bb.png)

![image](https://user-images.githubusercontent.com/7505593/88399569-f9ffe980-cdbe-11ea-96c4-0785b03fd4b8.png)

![image](https://user-images.githubusercontent.com/7505593/88399584-01bf8e00-cdbf-11ea-9a49-5ad8015b36c3.png)

## Follow-up stories

This work will be continued with the costs display stories coming shortly.
